proposals/js-promise-integration/Overview.md

```diff
@@ -2,299 +2,225 @@
+
+>Note that *`func`* is expected to be a *JavaScript function*. This allows us to ignore certain so-called corner cases in the usage of JSPI: in particular there is no special handling of WebAssembly functions passed to `WebAssembly.Suspending`.
+
+1. If IsCallable(*func*) is `false`, throw a `TypeError` exception.
+1. Let *suspendingProto* be `WebAssembly.Suspendable.%prototype%`
+1. Let *susp* be the result of `OrdinaryObjectCreate`(*`suspendingProto`*)
+1. Assign the `[[wrappedFunction]]` internal slot of *`susp`* to *`func`*
+1. Return *susp*
+
+The most direct way that external functions can be accessed from a WebAssembly module is via imports.[^WebAssembly.Function] Therefore, we modify the *read-the-imports* algorithm to account for imports annotated as `Suspendable`.
+
+[^WebAssembly.Function]: The other way is by using a `WebAssembly.Function` constructor, as proposed in the js-types proposal. However, the semantics of `WebAssembly.Function` also revolve around modules and importing.
 
-1. **What is the purpose of the `Suspender` object?**
+We replace the section dealing with functions:
+
+3.4. If *externtype* is of the form `func` *functype* ...
+
+with:
+
+3.4. If *externtype* is of the form `func` *functype*
+
+  1. If `$IsCallable$`(*`v`*) is `true`
+      1. If *`v`* has a *`[[FunctionAddress]]`* internal slot, and therefore is an *Exported Function*,
+          * Let *`funcaddr`* be the value of *`v`*'s *`[[FunctionAddress]]`* internal slot.
+      1. Otherwise,
+          * *Create a host function* from *`v`* and *`functype`*, and let *`funcaddr`* be the result.
+  1. If *`v`* has a *`[[wrappedFunction]]`* internal slot:
+      1. Let *`func`* be *`v`*'s *`[[wrappedFunction]]`* slot.
+      1. Assert `$IsCallable$`(*`func`*).
+      1. Create a *suspending function* from *`func`* and *`functype`*, and let *`funcaddr`* be the result.
+  1. If `$IsCallable$`(*`v`*) is `false` and *`v`* does not have a *`[[wrappedFunction]]`* internal slot then throw a `LinkError` exception.
+  1. Let `*index*` be the number of external functions in *`imports`*. This value `*index*` is known as the *index of the host function* *`funcaddr`*.
+  1. Let *`externfunc`* be the external value *`funcaddr`*
+  1. Append *`externfunc`* to *`imports`*.
+
+The *suspending function* is a function whose behavior is determined as follows:
+
+1. Let `context` refer to the execution context that is current at the time of a call to the *suspending function*. Let `wfunc` be the wrapped function that was used when creating the *suspending function*.
+1. Traps if `context`'s state is not **Active**[`caller`] for some `caller`
+1. Let `result` be the result of calling `wfunc(args)` (or any trap or thrown exception) where `args` are the additional arguments passed to the call when the imported function was called from the WebAssembly module.
+1. If `$isPromise(result)`:
```

Comment:
What does $isPromise refer to?

Reply:
EmcaScript abstract function checking for the Promise brand. It is also not quite correctly spelt here.